# Visualization using Plotly

<div class="alert alert-block alert-info">
<strong>Questions:</strong>

<ul>
    <li>How do I use Python to interactively visualize data?</li>
</ul>

<strong>Objectives:</strong>

<ul>
    <li>Use plotly to visualize data</li>
</ul>
</div>

Plotly is a Python library that allows you to create interactive visualizations. 
For this demonstration, we'll be using Plotly Express, a simple interface to the capabilities of Plotly.

We will first read in our data using pandas, then we will plot data using plotly. 
Plotly express is imported as `import plotly.express as px`

**Instructor edition.** Completed examples and worked solutions. Run all cells from top to bottom using the project Python environment. Data are read from the parent repository.

In [1]:
from pathlib import Path
import sys

# Support kernels started in the repository root or in instructor-notebooks.
ROOT = Path.cwd()
if not (ROOT / "data").is_dir():
    ROOT = ROOT.parent
if not (ROOT / "data").is_dir():
    raise FileNotFoundError("Open this notebook from the repository or instructor-notebooks folder.")
sys.path.insert(0, str(ROOT))

In [2]:
import pandas as pd

import plotly.express as px

In [3]:
df = pd.read_csv(ROOT / "data/PubChemElements_all.csv")
df.head()

,AtomicNumber,Symbol,Name,AtomicMass,CPKHexColor,ElectronConfiguration,Electronegativity,AtomicRadius,IonizationEnergy,ElectronAffinity,OxidationStates,StandardState,MeltingPoint,BoilingPoint,Density,GroupBlock,YearDiscovered
0,1,H,Hydrogen,1.008000,FFFFFF,1s1,2.20,120.0,13.598,0.754,"+1, -1",Gas,13.81,20.28,0.000090,Nonmetal,1766
1,2,He,Helium,4.002600,D9FFFF,1s2,NaN,140.0,24.587,NaN,0,Gas,0.95,4.22,0.000179,Noble gas,1868
2,3,Li,Lithium,7.000000,CC80FF,[He]2s1,0.98,182.0,5.392,0.618,+1,Solid,453.65,1615.00,0.534000,Alkali metal,1817
3,4,Be,Beryllium,9.012183,C2FF00,[He]2s2,1.57,153.0,9.323,NaN,+2,Solid,1560.00,2744.00,1.850000,Alkaline earth metal,1798
4,5,B,Boron,10.810000,FFB5B5,[He]2s2 2p1,2.04,192.0,8.298,0.277,+3,Solid,2348.00,4273.00,2.370000,Metalloid,1808


To plot something using plotly, we can use `px.line` or `px.scatter` for line and scatter plots respectively.

In [4]:
fig = px.line(df, x="AtomicNumber", y="MeltingPoint")
fig.add_scatter(x=df["AtomicNumber"], y=df["BoilingPoint"], name="Boiling Point")
fig.show()

Once you have a variable representing a figure, you can continue to update it in different cells.
This figure includes both melting and boiling points, so we will label the y axis "Temperature (K)".


In [5]:
fig.update_yaxes(title="Temperature (K)")

In [6]:
fig.show()

Plotly allows you to set the color of your markers based on the value in another column. In the cell below, we color the points based on their block.

In [7]:
fig = px.scatter(df, x="AtomicNumber", y="IonizationEnergy", color="GroupBlock")
fig.show()

We can also color elements by their period (row) in the periodic table.
The example below uses [pandas `cut`](https://pandas.pydata.org/docs/reference/api/pandas.cut.html) to group atomic numbers into intervals.
Each interval ends at the last atomic number in a period: 2, 10, 18, 36, 54, 86, and 118.
For example, atomic numbers 3 through 10 receive the label "2".
We save these labels in a column called `Period`.

This is an optional data-preparation example; you are not expected to write it from memory.
The following plot uses `color="Period"` to color by those labels and `hover_data=["Name"]` to show element names when you hover over points.


In [8]:
df["Period"] = pd.cut(
    df["AtomicNumber"],
    bins=[0, 2, 10, 18, 36, 54, 86, 118],
    labels=["1", "2", "3", "4", "5", "6", "7"],
)


In [9]:
fig = px.scatter(df, x="AtomicNumber", y="IonizationEnergy", color="Period", hover_data=["Name"])
fig.show()


In [10]:
df[["Name", "AtomicNumber", "Period", "IonizationEnergy"]]

,Name,AtomicNumber,Period,IonizationEnergy
0,Hydrogen,1,1,13.598
1,Helium,2,1,24.587
2,Lithium,3,2,5.392
3,Beryllium,4,2,9.323
4,Boron,5,2,8.298
...,...,...,...,...
113,Flerovium,114,7,NaN
114,Moscovium,115,7,NaN
115,Livermorium,116,7,NaN
116,Tennessine,117,7,NaN


<div class="alert alert-block alert-warning">
<strong>Check Your Understanding</strong>

<p>
Create a plot of <code>AtomicRadius</code> vs <code>AtomicNumber</code>.
Update your axes to have appropriate labels.
Try coloring your plot by <code>GroupBlock</code>, then by <code>Period</code>.
</p>

<p>
What other periodic trends can you explore with visualization?
</p>
</div>

In [11]:
fig = px.scatter(df, x="AtomicNumber", y="AtomicRadius", color="GroupBlock", symbol="GroupBlock",
                 hover_data=["Name"], title="Atomic radius across the periodic table",
                 labels={"AtomicNumber": "Atomic number", "AtomicRadius": "Atomic radius (pm)", "GroupBlock": "Element group"})
fig.update_traces(marker=dict(size=9, line=dict(width=1, color="#222222")))
fig.show()

In [12]:
fig = px.scatter(df, x="AtomicNumber", y="AtomicRadius", color="Period", symbol="Period",
                 hover_data=["Name"], title="Atomic radius by period",
                 labels={"AtomicNumber": "Atomic number", "AtomicRadius": "Atomic radius (pm)", "GroupBlock": "Element group"})
fig.update_traces(marker=dict(size=9, line=dict(width=1, color="#222222")))
fig.show()

In [13]:
df[["Name", "AtomicNumber", "AtomicRadius", "GroupBlock", "Period"]]

,Name,AtomicNumber,AtomicRadius,GroupBlock,Period
0,Hydrogen,1,120.0,Nonmetal,1
1,Helium,2,140.0,Noble gas,1
2,Lithium,3,182.0,Alkali metal,2
3,Beryllium,4,153.0,Alkaline earth metal,2
4,Boron,5,192.0,Metalloid,2
...,...,...,...,...,...
113,Flerovium,114,180.0,Post-transition metal,7
114,Moscovium,115,187.0,Post-transition metal,7
115,Livermorium,116,183.0,Post-transition metal,7
116,Tennessine,117,138.0,Halogen,7


In [14]:
fig = px.scatter(df, x="AtomicNumber", y="Electronegativity", color="GroupBlock", symbol="GroupBlock", hover_data=["Name"], title="Electronegativity across the periodic table", labels={"AtomicNumber": "Atomic number", "Electronegativity": "Electronegativity (Pauling scale)"})
fig.show()

We can also do thing like visualize boiling points by standard states using a box plot.

In [15]:
fig = px.box(df, x="StandardState", y="BoilingPoint")
fig.show()

In [16]:
df.groupby("StandardState")["BoilingPoint"].describe()

,count,mean,std,min,25%,50%,75%,max
StandardState,,,,,,,,
Expected to be a Gas,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Expected to be a Solid,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Gas,11.0,102.452727,76.272209,4.22,52.2150,87.300,142.4800,239.11
Liquid,2.0,480.915000,210.668323,331.95,406.4325,480.915,555.3975,629.88
Solid,80.0,2922.236875,1361.743303,457.55,1790.7500,3107.500,3707.0000,5869.00


You can see an overview of available plot types in plotly express [on this page](https://plotly.com/python/plotly-express/).

In [17]:
fig = px.box(df, x="Period", y="IonizationEnergy")
fig.show()


<div class="alert alert-block alert-warning">
<strong>Final Exercise: Create Your Own Notebook</strong>

<p>
<strong>Create a new notebook</strong> named <code>03-periodic-trends-exercise.ipynb</code> in the same folder as the lesson notebooks. Complete this exercise there.
</p>
<p>
Start with a Markdown cell containing a title and a brief description of the problem. Add code cells for your setup and analysis. Import pandas and Plotly Express, then load <code>data/PubChemElements_all.csv</code> into a new DataFrame.
</p>
<p>
Each notebook needs its own setup: variables and imports from the lesson notebook are not automatically available in your new notebook.
</p>

<ol>
    <li>Create a scatter plot of <code>AtomicRadius</code> against <code>AtomicNumber</code>, with points colored by <code>GroupBlock</code>.</li>
    <li>Give the plot a descriptive title and clear axis labels, and include each element's <code>Name</code> in the hover information.</li>
    <li>Create a box plot comparing <code>BoilingPoint</code> across <code>StandardState</code> categories.</li>
    <li>Describe one pattern you observe in each plot. What does each plot help you compare?</li>
</ol>

<p>
Finish with a Markdown cell explaining your results. Save your notebook, restart its kernel, and run all cells from top to bottom. Check that it completes without errors and reproduces your results. This checks that every step needed for your analysis is included in the notebook.
</p>
</div>


**Worked solution:** [03-periodic-trends-exercise.ipynb](03-periodic-trends-exercise.ipynb). This separate notebook includes all required setup and interpretation.

**Instructor interpretation:** Atomic radius shows repeating variation across atomic number, generally decreasing across a period and increasing down a group. A scatter plot preserves the individual element measurements. The box plot compares distributions: median boiling points are 87.3 K for gases, 480.915 K for liquids, and 3107.5 K for solids in this dataset. Missing measurements are omitted, including categories containing only missing values. Use the tables below for a text alternative to the plots.

In [18]:
df[["Name", "AtomicNumber", "AtomicRadius", "IonizationEnergy", "Electronegativity", "GroupBlock", "Period", "StandardState", "BoilingPoint"]]

,Name,AtomicNumber,AtomicRadius,IonizationEnergy,Electronegativity,GroupBlock,Period,StandardState,BoilingPoint
0,Hydrogen,1,120.0,13.598,2.20,Nonmetal,1,Gas,20.28
1,Helium,2,140.0,24.587,NaN,Noble gas,1,Gas,4.22
2,Lithium,3,182.0,5.392,0.98,Alkali metal,2,Solid,1615.00
3,Beryllium,4,153.0,9.323,1.57,Alkaline earth metal,2,Solid,2744.00
4,Boron,5,192.0,8.298,2.04,Metalloid,2,Solid,4273.00
...,...,...,...,...,...,...,...,...,...
113,Flerovium,114,180.0,NaN,NaN,Post-transition metal,7,Expected to be a Solid,NaN
114,Moscovium,115,187.0,NaN,NaN,Post-transition metal,7,Expected to be a Solid,NaN
115,Livermorium,116,183.0,NaN,NaN,Post-transition metal,7,Expected to be a Solid,NaN
116,Tennessine,117,138.0,NaN,NaN,Halogen,7,Expected to be a Solid,NaN
